---
title: Contact Us
layout: default
active: support
---
# Citation and Contact

The MIMOSA2 manuscript is currently in preparation. If you use MIMOSA2 in a publication, please include the URL **[www.borensteinlab.com/software_MIMOSA2.html](http://borensteinlab.com/software_MIMOSA2.html)**, and cite the publication describing MIMOSA v1:

Noecker, C., Eng, A., Srinivasan, S., Theriot, C.M., Young, V.B., Jansson, J.K., Fredricks, D.N., and Borenstein, E. (2016). Metabolic Model-Based Integration of Microbiome Taxonomic and Metabolomic Profiles Elucidates Mechanistic Links between Ecological and Metabolic Variation. mSystems 1, e00013–15, [doi:10.1128/mSystems.00013-15](http://doi.org/10.1128/mSystems.00013-15).

MIMOSA2 uses several other R packages, bioinformatics tools, and databases. Depending on the settings for your MIMOSA2 analysis, you may also need to cite the following:

- [vsearch](https://github.com/torognes/vsearch): Rognes, T., Flouri, T., Nichols, B., Quince, C., and Mahé, F. (2016). VSEARCH: a versatile open source tool for metagenomics. PeerJ 4, e2584.

- [PICRUSt](https://picrust.github.io/picrust/): Langille, M.G.I., Zaneveld, J., Caporaso, J.G., McDonald, D., Knights, D., Reyes, J.A., Clemente, J.C., Burkepile, D.E., Vega Thurber, R.L., Knight, R., et al. (2013). Predictive functional profiling of microbial communities using 16S rRNA marker gene sequences. Nat Biotech 31, 814–821.

- [MetaboAnalystR](https://www.metaboanalyst.ca): Chong, J., Soufan, O., Li, C., Caraus, I., Li, S., Bourque, G., Wishart, D.S., and Xia, J. (2018). MetaboAnalyst 4.0: towards more transparent and integrative metabolomics analysis. Nucleic Acids Research.

- [AGORA](vmh.uni.lu): Magnúsdóttir, S., Heinken, A., Kutt, L., Ravcheev, D.A., Bauer, E., Noronha, A., Greenhalgh, K., Jäger, C., Baginska, J., Wilmes, P., et al. (2016). Generation of genome-scale metabolic reconstructions for 773 members of the human gut microbiota. Nature Biotechnology.

- [CarveMe/embl_gems](https://github.com/cdanielmachado/embl_gems): Machado, D., Andrejev, S., Tramontano, M., and Patil, K.R. (2018). Fast automated reconstruction of genome-scale metabolic models for microbial species and communities. Nucleic Acids Research 46, 7542–7553.

- [KEGG](https://www.genome.jp/kegg/): Kanehisa, M., and Goto, S. (2000). KEGG: kyoto encyclopedia of genes and genomes. Nucleic Acids Research 28, 27–30.

- [Rfit](https://journal.r-project.org/archive/2012-2/RJournal_2012-2_Kloke+McKean.pdf): Kloke, J.D. and Mckean, J.W. (2012), Rfit: Rank-based estimation for linear models, The R Journal, 4, 57--64.

For questions about MIMOSA2, please email elbo [at] tauex.tau.ac.il  .

<!---
Noecker, C. and E. Borenstein. MIMOSA2: A metabolic network-based tool for linking microbiome and metabolomics data. In prep, 2019.
-->